File: app/src/main/java/com/example/flashcardapp/TopicSelectionAdapter.java
// File: TopicSelectionAdapter.java
package com.example.flashcardapp;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import java.util.List;

public class TopicSelectionAdapter extends RecyclerView.Adapter<TopicSelectionAdapter.TopicViewHolder> {

    private final List<Topic> topics;

    public TopicSelectionAdapter(List<Topic> topics) {
        this.topics = topics;
    }

    @NonNull
    @Override
    public TopicViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_topic_selection, parent, false);
        return new TopicViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull TopicViewHolder holder, int position) {
        Topic topic = topics.get(position);
        holder.cbTopic.setText(topic.getName());
        holder.cbTopic.setChecked(topic.isSelected());

        // Update selection state on checkbox change
        holder.cbTopic.setOnCheckedChangeListener((buttonView, isChecked) -> topic.setSelected(isChecked));
    }

    @Override
    public int getItemCount() {
        return topics.size();
    }

    public List<Topic> getSelectedTopics() {
        return topics;
    }

    static class TopicViewHolder extends RecyclerView.ViewHolder {
        CheckBox cbTopic;

        public TopicViewHolder(@NonNull View itemView) {
            super(itemView);
            cbTopic = itemView.findViewById(R.id.cb_topic);
        }
    }
}

File: app/src/main/java/com/example/flashcardapp/main/TopicSelectionAdapter.kt
package com.example.flashcardapp.main

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import androidx.recyclerview.widget.RecyclerView
import com.example.flashcardapp.R
import com.example.flashcardapp.data.TopicWithCount

// changes: now we store TopicWithCount
class TopicSelectionAdapter(private val topics: List<TopicWithCount>) :
    RecyclerView.Adapter<TopicSelectionAdapter.TopicViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TopicViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_topic_selection, parent, false)
        return TopicViewHolder(view)
    }

    override fun onBindViewHolder(holder: TopicViewHolder, position: Int) {
        val topicWC = topics[position]
        // show count next to the name
        val label = "${topicWC.name} (${topicWC.cardCount})"

        holder.cbTopic.text = label
        holder.cbTopic.isChecked = topicWC.selected
        holder.cbTopic.setOnCheckedChangeListener { _, isChecked ->
            topicWC.selected = isChecked
        }
    }

    override fun getItemCount(): Int {
        return topics.size
    }

    // Return the list (for saving) after user toggles or individually checks/unchecks.
    fun getSelectedTopics(): List<TopicWithCount> {
        return topics
    }

    // NEW: Check if everything is selected
    fun areAllSelected(): Boolean {
        return topics.all { it.selected }
    }

    // NEW: Select all
    fun selectAll() {
        for (t in topics) {
            t.selected = true
        }
        notifyDataSetChanged()
    }

    // NEW: Deselect all
    fun deselectAll() {
        for (t in topics) {
            t.selected = false
        }
        notifyDataSetChanged()
    }

    class TopicViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val cbTopic: CheckBox = itemView.findViewById(R.id.cb_topic)
    }
}
